{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Asynchronous queries on BioMart using `apybiomart` \n"
   ],
   "metadata": {
    "pycharm": {
     "metadata": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "[BioMart](https://www.ensembl.org/biomart/martview) is a web-based tool that allows to retrieve and export datasets from Ensembl. It is one of the most useful resources for bioinformatics, because it makes easy to retrieve data and further analyse them or integrate them in downstream analyses. For these reasons, researchers have developed applications to perform BioMart queries from within R and Python programs, namely [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html) and [pybiomart](https://github.com/jrderuiter/pybiomart).  \n",
    "\n",
    "I've been using both for different needs, and I was very happy with them. But lately I've been working on a database that needs to access many different online resources, and thus performs asynchronous requests (exploiting the great `asyncio` Python library). Since the database was developed in Python, I tried to use `pybiomart` for these tasks, but unfortunately it was not suitable to perform async calls, so I decided to modify it to better suit my needs, and that's when `apybiomart` was born!  "
   ],
   "metadata": {
    "pycharm": {
     "metadata": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "What this package does is simply modifying the original `pybiomart` behaviour to retain its basic capabilities, extending them to handle asynchronous requests to the BioMart service. As such, it is possible to perform multiple queries at the same time.  "
   ],
   "metadata": {
    "pycharm": {
     "metadata": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Installation  \n",
    "Apybiomart **only supports Python 3**, and can be installed through `pip`:  \n",
    "```bash\n",
    "pip install apybiomart\n",
    "```"
   ],
   "metadata": {
    "pycharm": {
     "metadata": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Usage  \n",
    "After installation, apybiomart can be easily imported into Python:  "
   ],
   "metadata": {
    "pycharm": {
     "metadata": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "outputs": [],
   "source": [
    "import apybiomart"
   ],
   "metadata": {
    "pycharm": {
     "metadata": false,
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "Now it is possible to use its functions:  \n",
    "\n",
    "* explore the list of available marts, datasets, attributes and filters using the synchronous `find_*()` functions;  \n",
    "* perform a single synchronous query using the `query()` function;  \n",
    "* perform multiple asynchronous queries using the `aquery()` function.  "
   ],
   "metadata": {
    "pycharm": {
     "metadata": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Marts, datasets, attributes and filters  \n",
    "\n",
    "BioMart contains different databases, called *marts*, each of which in turn contains several *datasets*, each related to a particular species. These datasets can be queried and it is possible to restrict the amount of data returned to one or more particular types of information, namely *attributes*, and using *filters* that only retain data satisfying one or more specific criteria (more information on BioMart's [help page](https://www.ensembl.org/info/data/biomart/index.html).  \n",
    "\n",
    "This tutorial will mostly follow the outline available in the official [biomaRt users guide](https://www.bioconductor.org/packages/devel/bioc/vignettes/biomaRt/inst/doc/biomaRt.html).  "
   ],
   "metadata": {
    "pycharm": {
     "metadata": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### Marts  \n",
    "\n",
    "In order to view marts available on BioMart, use `find_marts()`; a dataframe with the available marts is returned, with their proper `name` and `display_name`:  "
   ],
   "metadata": {
    "pycharm": {
     "metadata": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [
    {
     "data": {
      "text/plain": "                    name           display_name\n0   ENSEMBL_MART_ENSEMBL       Ensembl Genes 96\n1     ENSEMBL_MART_MOUSE       Mouse strains 96\n2  ENSEMBL_MART_SEQUENCE               Sequence\n3  ENSEMBL_MART_ONTOLOGY               Ontology\n4   ENSEMBL_MART_GENOMIC    Genomic features 96\n5       ENSEMBL_MART_SNP   Ensembl Variation 96\n6   ENSEMBL_MART_FUNCGEN  Ensembl Regulation 96",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>name</th>\n      <th>display_name</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n      <td>Ensembl Genes 96</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>ENSEMBL_MART_MOUSE</td>\n      <td>Mouse strains 96</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENSEMBL_MART_SEQUENCE</td>\n      <td>Sequence</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENSEMBL_MART_ONTOLOGY</td>\n      <td>Ontology</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENSEMBL_MART_GENOMIC</td>\n      <td>Genomic features 96</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ENSEMBL_MART_SNP</td>\n      <td>Ensembl Variation 96</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>ENSEMBL_MART_FUNCGEN</td>\n      <td>Ensembl Regulation 96</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "metadata": {},
     "output_type": "execute_result",
     "execution_count": 2
    }
   ],
   "source": [
    "apybiomart.find_marts()"
   ],
   "metadata": {
    "pycharm": {
     "metadata": false,
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### Datasets  \n",
    "\n",
    "Each BioMart database (or mart) can contain several different datasets, each referring to a particular species. The `find_datasets()` function allows to list all the available datasets for a given mart, which by default is \"ENSEMBL_MART_ENSEMBL\".  "
   ],
   "metadata": {
    "pycharm": {
     "metadata": false,
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [
    {
     "data": {
      "text/plain": "                           name  \\\n0      uamericanus_gene_ensembl   \n1          ngalili_gene_ensembl   \n2        oprinceps_gene_ensembl   \n3        oanatinus_gene_ensembl   \n4           malbus_gene_ensembl   \n5      mnemestrina_gene_ensembl   \n6            catys_gene_ensembl   \n7          elucius_gene_ensembl   \n8           drerio_gene_ensembl   \n9      zalbicollis_gene_ensembl   \n10       lcoronata_gene_ensembl   \n11         cpugnax_gene_ensembl   \n12          bbison_gene_ensembl   \n13      mdomestica_gene_ensembl   \n14        jjaculus_gene_ensembl   \n15        hburtoni_gene_ensembl   \n16    acitrinellus_gene_ensembl   \n17     fdamarensis_gene_ensembl   \n18      cporcellus_gene_ensembl   \n19        olatipes_gene_ensembl   \n20      cchok1gshd_gene_ensembl   \n21      oniloticus_gene_ensembl   \n22        marmatus_gene_ensembl   \n23    mleucophaeus_gene_ensembl   \n24      xmaculatus_gene_ensembl   \n25           arowi_gene_ensembl   \n26      umaritimus_gene_ensembl   \n27           hmale_gene_ensembl   \n28       mmusculus_gene_ensembl   \n29     rnorvegicus_gene_ensembl   \n..                          ...   \n154     ipunctatus_gene_ensembl   \n155         dordii_gene_ensembl   \n156         pmajor_gene_ensembl   \n157         mzebra_gene_ensembl   \n158     choffmanni_gene_ensembl   \n159     pcoquereli_gene_ensembl   \n160       hsapiens_gene_ensembl   \n161       celegans_gene_ensembl   \n162   sboliviensis_gene_ensembl   \n163       apercula_gene_ensembl   \n164      etelfairi_gene_ensembl   \n165    mspicilegus_gene_ensembl   \n166        chircus_gene_ensembl   \n167     rroxellana_gene_ensembl   \n168        panubis_gene_ensembl   \n169        mcaroli_gene_ensembl   \n170      clanigera_gene_ensembl   \n171   atestudineus_gene_ensembl   \n172         vpacos_gene_ensembl   \n173      pcapensis_gene_ensembl   \n174      spartitus_gene_ensembl   \n175     ttruncatus_gene_ensembl   \n176   mochrogaster_gene_ensembl   \n177  acarolinensis_gene_ensembl   \n178       smaximus_gene_ensembl   \n179       tguttata_gene_ensembl   \n180       cpygmaea_gene_ensembl   \n181       neugenii_gene_ensembl   \n182     pnattereri_gene_ensembl   \n183     ccapucinus_gene_ensembl   \n\n                                          display_name                  mart  \n0              American black bear genes (ASM334442v1)  ENSEMBL_MART_ENSEMBL  \n1    Upper Galilee mountains blind mole rat genes (...  ENSEMBL_MART_ENSEMBL  \n2                           Pika genes (OchPri2.0-Ens)  ENSEMBL_MART_ENSEMBL  \n3                               Platypus genes (OANA5)  ENSEMBL_MART_ENSEMBL  \n4                        Swamp eel genes (M_albus_1.0)  ENSEMBL_MART_ENSEMBL  \n5                  Pig-tailed macaque genes (Mnem_1.0)  ENSEMBL_MART_ENSEMBL  \n6                      Sooty mangabey genes (Caty_1.0)  ENSEMBL_MART_ENSEMBL  \n7                        Northern pike genes (Eluc_V3)  ENSEMBL_MART_ENSEMBL  \n8                             Zebrafish genes (GRCz11)  ENSEMBL_MART_ENSEMBL  \n9    White-throated sparrow genes (Zonotrichia_albi...  ENSEMBL_MART_ENSEMBL  \n10   Blue-crowned manakin genes (Lepidothrix_corona...  ENSEMBL_MART_ENSEMBL  \n11                            Ruff genes (ASM143184v1)  ENSEMBL_MART_ENSEMBL  \n12                 American bison genes (Bison_UMD1.0)  ENSEMBL_MART_ENSEMBL  \n13                             Opossum genes (monDom5)  ENSEMBL_MART_ENSEMBL  \n14            Lesser Egyptian jerboa genes (JacJac1.0)  ENSEMBL_MART_ENSEMBL  \n15             Burton's mouthbrooder genes (AstBur1.0)  ENSEMBL_MART_ENSEMBL  \n16                      Midas cichlid genes (Midas_v5)  ENSEMBL_MART_ENSEMBL  \n17                    Damara mole rat genes (DMR_v1.0)  ENSEMBL_MART_ENSEMBL  \n18                        Guinea Pig genes (Cavpor3.0)  ENSEMBL_MART_ENSEMBL  \n19            Japanese medaka HdrR genes (ASM223467v1)  ENSEMBL_MART_ENSEMBL  \n20        Chinese hamster CHOK1GS genes (CHOK1GS_HDv1)  ENSEMBL_MART_ENSEMBL  \n21                           Tilapia genes (Orenil1.0)  ENSEMBL_MART_ENSEMBL  \n22                      Zig-zag eel genes (fMasArm1.1)  ENSEMBL_MART_ENSEMBL  \n23                           Drill genes (Mleu.le_1.0)  ENSEMBL_MART_ENSEMBL  \n24              Platyfish genes (X_maculatus-5.0-male)  ENSEMBL_MART_ENSEMBL  \n25                  Okarito brown kiwi genes (aptRow1)  ENSEMBL_MART_ENSEMBL  \n26                       Polar bear genes (UrsMar_1.0)  ENSEMBL_MART_ENSEMBL  \n27              Naked mole-rat male genes (HetGla_1.0)  ENSEMBL_MART_ENSEMBL  \n28                             Mouse genes (GRCm38.p6)  ENSEMBL_MART_ENSEMBL  \n29                                Rat genes (Rnor_6.0)  ENSEMBL_MART_ENSEMBL  \n..                                                 ...                   ...  \n154                 Channel catfish genes (IpCoco_1.2)  ENSEMBL_MART_ENSEMBL  \n155                      Kangaroo rat genes (Dord_2.0)  ENSEMBL_MART_ENSEMBL  \n156                   Great Tit genes (Parus_major1.1)  ENSEMBL_MART_ENSEMBL  \n157                  Zebra mbuna genes (M_zebra_UMD2a)  ENSEMBL_MART_ENSEMBL  \n158                              Sloth genes (choHof1)  ENSEMBL_MART_ENSEMBL  \n159                 Coquerel's sifaka genes (Pcoq_1.0)  ENSEMBL_MART_ENSEMBL  \n160                           Human genes (GRCh38.p12)  ENSEMBL_MART_ENSEMBL  \n161            Caenorhabditis elegans genes (WBcel235)  ENSEMBL_MART_ENSEMBL  \n162         Bolivian squirrel monkey genes (SaiBol1.0)  ENSEMBL_MART_ENSEMBL  \n163                   Orange clownfish genes (Nemo_v1)  ENSEMBL_MART_ENSEMBL  \n164              Lesser hedgehog tenrec genes (TENREC)  ENSEMBL_MART_ENSEMBL  \n165                       Steppe mouse genes (MUSP714)  ENSEMBL_MART_ENSEMBL  \n166                                  Goat genes (ARS1)  ENSEMBL_MART_ENSEMBL  \n167           Golden snub-nosed monkey genes (Rrox_v1)  ENSEMBL_MART_ENSEMBL  \n168                      Olive baboon genes (Panu_3.0)  ENSEMBL_MART_ENSEMBL  \n169               Ryukyu mouse genes (CAROLI_EIJ_v1.1)  ENSEMBL_MART_ENSEMBL  \n170           Long-tailed chinchilla genes (ChiLan1.0)  ENSEMBL_MART_ENSEMBL  \n171                  Climbing perch genes (fAnaTes1.1)  ENSEMBL_MART_ENSEMBL  \n172                             Alpaca genes (vicPac1)  ENSEMBL_MART_ENSEMBL  \n173                              Hyrax genes (proCap1)  ENSEMBL_MART_ENSEMBL  \n174  Bicolor damselfish genes (Stegastes_partitus-1...  ENSEMBL_MART_ENSEMBL  \n175                            Dolphin genes (turTru1)  ENSEMBL_MART_ENSEMBL  \n176                     Prairie vole genes (MicOch1.0)  ENSEMBL_MART_ENSEMBL  \n177                     Anole lizard genes (AnoCar2.0)  ENSEMBL_MART_ENSEMBL  \n178                         Turbot genes (ASM318616v1)  ENSEMBL_MART_ENSEMBL  \n179                    Zebra Finch genes (taeGut3.2.4)  ENSEMBL_MART_ENSEMBL  \n180         Spoon-billed sandpiper genes (ASM369795v1)  ENSEMBL_MART_ENSEMBL  \n181                           Wallaby genes (Meug_1.0)  ENSEMBL_MART_ENSEMBL  \n182  Red-bellied piranha genes (Pygocentrus_nattere...  ENSEMBL_MART_ENSEMBL  \n183                Capuchin genes (Cebus_imitator-1.0)  ENSEMBL_MART_ENSEMBL  \n\n[184 rows x 3 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>name</th>\n      <th>display_name</th>\n      <th>mart</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>uamericanus_gene_ensembl</td>\n      <td>American black bear genes (ASM334442v1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>ngalili_gene_ensembl</td>\n      <td>Upper Galilee mountains blind mole rat genes (...</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>oprinceps_gene_ensembl</td>\n      <td>Pika genes (OchPri2.0-Ens)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>oanatinus_gene_ensembl</td>\n      <td>Platypus genes (OANA5)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>malbus_gene_ensembl</td>\n      <td>Swamp eel genes (M_albus_1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>mnemestrina_gene_ensembl</td>\n      <td>Pig-tailed macaque genes (Mnem_1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>catys_gene_ensembl</td>\n      <td>Sooty mangabey genes (Caty_1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>elucius_gene_ensembl</td>\n      <td>Northern pike genes (Eluc_V3)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>drerio_gene_ensembl</td>\n      <td>Zebrafish genes (GRCz11)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>zalbicollis_gene_ensembl</td>\n      <td>White-throated sparrow genes (Zonotrichia_albi...</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>lcoronata_gene_ensembl</td>\n      <td>Blue-crowned manakin genes (Lepidothrix_corona...</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>cpugnax_gene_ensembl</td>\n      <td>Ruff genes (ASM143184v1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>bbison_gene_ensembl</td>\n      <td>American bison genes (Bison_UMD1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>mdomestica_gene_ensembl</td>\n      <td>Opossum genes (monDom5)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>jjaculus_gene_ensembl</td>\n      <td>Lesser Egyptian jerboa genes (JacJac1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>hburtoni_gene_ensembl</td>\n      <td>Burton's mouthbrooder genes (AstBur1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>acitrinellus_gene_ensembl</td>\n      <td>Midas cichlid genes (Midas_v5)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>fdamarensis_gene_ensembl</td>\n      <td>Damara mole rat genes (DMR_v1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>cporcellus_gene_ensembl</td>\n      <td>Guinea Pig genes (Cavpor3.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>olatipes_gene_ensembl</td>\n      <td>Japanese medaka HdrR genes (ASM223467v1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>cchok1gshd_gene_ensembl</td>\n      <td>Chinese hamster CHOK1GS genes (CHOK1GS_HDv1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>oniloticus_gene_ensembl</td>\n      <td>Tilapia genes (Orenil1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>marmatus_gene_ensembl</td>\n      <td>Zig-zag eel genes (fMasArm1.1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>mleucophaeus_gene_ensembl</td>\n      <td>Drill genes (Mleu.le_1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>xmaculatus_gene_ensembl</td>\n      <td>Platyfish genes (X_maculatus-5.0-male)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>arowi_gene_ensembl</td>\n      <td>Okarito brown kiwi genes (aptRow1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>26</th>\n      <td>umaritimus_gene_ensembl</td>\n      <td>Polar bear genes (UrsMar_1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>27</th>\n      <td>hmale_gene_ensembl</td>\n      <td>Naked mole-rat male genes (HetGla_1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>28</th>\n      <td>mmusculus_gene_ensembl</td>\n      <td>Mouse genes (GRCm38.p6)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>29</th>\n      <td>rnorvegicus_gene_ensembl</td>\n      <td>Rat genes (Rnor_6.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>154</th>\n      <td>ipunctatus_gene_ensembl</td>\n      <td>Channel catfish genes (IpCoco_1.2)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>155</th>\n      <td>dordii_gene_ensembl</td>\n      <td>Kangaroo rat genes (Dord_2.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>156</th>\n      <td>pmajor_gene_ensembl</td>\n      <td>Great Tit genes (Parus_major1.1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>157</th>\n      <td>mzebra_gene_ensembl</td>\n      <td>Zebra mbuna genes (M_zebra_UMD2a)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>158</th>\n      <td>choffmanni_gene_ensembl</td>\n      <td>Sloth genes (choHof1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>159</th>\n      <td>pcoquereli_gene_ensembl</td>\n      <td>Coquerel's sifaka genes (Pcoq_1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>160</th>\n      <td>hsapiens_gene_ensembl</td>\n      <td>Human genes (GRCh38.p12)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>161</th>\n      <td>celegans_gene_ensembl</td>\n      <td>Caenorhabditis elegans genes (WBcel235)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>162</th>\n      <td>sboliviensis_gene_ensembl</td>\n      <td>Bolivian squirrel monkey genes (SaiBol1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>163</th>\n      <td>apercula_gene_ensembl</td>\n      <td>Orange clownfish genes (Nemo_v1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>164</th>\n      <td>etelfairi_gene_ensembl</td>\n      <td>Lesser hedgehog tenrec genes (TENREC)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>165</th>\n      <td>mspicilegus_gene_ensembl</td>\n      <td>Steppe mouse genes (MUSP714)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>166</th>\n      <td>chircus_gene_ensembl</td>\n      <td>Goat genes (ARS1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>167</th>\n      <td>rroxellana_gene_ensembl</td>\n      <td>Golden snub-nosed monkey genes (Rrox_v1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>168</th>\n      <td>panubis_gene_ensembl</td>\n      <td>Olive baboon genes (Panu_3.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>169</th>\n      <td>mcaroli_gene_ensembl</td>\n      <td>Ryukyu mouse genes (CAROLI_EIJ_v1.1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>170</th>\n      <td>clanigera_gene_ensembl</td>\n      <td>Long-tailed chinchilla genes (ChiLan1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>171</th>\n      <td>atestudineus_gene_ensembl</td>\n      <td>Climbing perch genes (fAnaTes1.1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>172</th>\n      <td>vpacos_gene_ensembl</td>\n      <td>Alpaca genes (vicPac1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>173</th>\n      <td>pcapensis_gene_ensembl</td>\n      <td>Hyrax genes (proCap1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>174</th>\n      <td>spartitus_gene_ensembl</td>\n      <td>Bicolor damselfish genes (Stegastes_partitus-1...</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>175</th>\n      <td>ttruncatus_gene_ensembl</td>\n      <td>Dolphin genes (turTru1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>176</th>\n      <td>mochrogaster_gene_ensembl</td>\n      <td>Prairie vole genes (MicOch1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>177</th>\n      <td>acarolinensis_gene_ensembl</td>\n      <td>Anole lizard genes (AnoCar2.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>178</th>\n      <td>smaximus_gene_ensembl</td>\n      <td>Turbot genes (ASM318616v1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>179</th>\n      <td>tguttata_gene_ensembl</td>\n      <td>Zebra Finch genes (taeGut3.2.4)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>180</th>\n      <td>cpygmaea_gene_ensembl</td>\n      <td>Spoon-billed sandpiper genes (ASM369795v1)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>181</th>\n      <td>neugenii_gene_ensembl</td>\n      <td>Wallaby genes (Meug_1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>182</th>\n      <td>pnattereri_gene_ensembl</td>\n      <td>Red-bellied piranha genes (Pygocentrus_nattere...</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n    <tr>\n      <th>183</th>\n      <td>ccapucinus_gene_ensembl</td>\n      <td>Capuchin genes (Cebus_imitator-1.0)</td>\n      <td>ENSEMBL_MART_ENSEMBL</td>\n    </tr>\n  </tbody>\n</table>\n<p>184 rows × 3 columns</p>\n</div>"
     },
     "metadata": {},
     "output_type": "execute_result",
     "execution_count": 2
    }
   ],
   "source": [
    "apybiomart.find_datasets()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### Attributes  \n",
    "\n",
    "When querying one of BioMart's datasets, attributes define specific values we might be interested in retrieving, such as gene symbols or chromosomal coordinates. The `find_attributes()` function displays all available attributes in a given dataset, which by default is \"hsapiens_gene_ensembl\":  "
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "data": {
      "text/plain": "                                   name  \\\n0                       ensembl_gene_id   \n1               ensembl_gene_id_version   \n2                 ensembl_transcript_id   \n3         ensembl_transcript_id_version   \n4                    ensembl_peptide_id   \n5            ensembl_peptide_id_version   \n6                       ensembl_exon_id   \n7                           description   \n8                       chromosome_name   \n9                        start_position   \n10                         end_position   \n11                               strand   \n12                                 band   \n13                     transcript_start   \n14                       transcript_end   \n15             transcription_start_site   \n16                    transcript_length   \n17                       transcript_tsl   \n18             transcript_gencode_basic   \n19                    transcript_appris   \n20               transcript_mane_select   \n21                   external_gene_name   \n22                 external_gene_source   \n23             external_transcript_name   \n24      external_transcript_source_name   \n25                     transcript_count   \n26           percentage_gene_gc_content   \n27                         gene_biotype   \n28                   transcript_biotype   \n29                               source   \n...                                 ...   \n3321        sequence_transcript_biotype   \n3322   sequence_transcript_chrom_strand   \n3323    sequence_transcript_chrom_start   \n3324      sequence_transcript_chrom_end   \n3325  sequence_transcription_start_site   \n3326            sequence_peptide_length   \n3327         sequence_transcript_length   \n3328               sequence_cdna_length   \n3329                  cdna_coding_start   \n3330                    cdna_coding_end   \n3331                        5_utr_start   \n3332                          5_utr_end   \n3333                        3_utr_start   \n3334                          3_utr_end   \n3335            sequence_exon_stable_id   \n3336                      sequence_type   \n3337          sequence_exon_chrom_start   \n3338            sequence_exon_chrom_end   \n3339         sequence_exon_chrom_strand   \n3340                      sequence_rank   \n3341                     sequence_phase   \n3342                 sequence_end_phase   \n3343         sequence_cdna_coding_start   \n3344           sequence_cdna_coding_end   \n3345      sequence_genomic_coding_start   \n3346        sequence_genomic_coding_end   \n3347              sequence_constitutive   \n3348                         cds_length   \n3349                          cds_start   \n3350                            cds_end   \n\n                                    display_name  \\\n0                                 Gene stable ID   \n1                         Gene stable ID version   \n2                           Transcript stable ID   \n3                   Transcript stable ID version   \n4                              Protein stable ID   \n5                      Protein stable ID version   \n6                                 Exon stable ID   \n7                               Gene description   \n8                       Chromosome/scaffold name   \n9                                Gene start (bp)   \n10                                 Gene end (bp)   \n11                                        Strand   \n12                                Karyotype band   \n13                         Transcript start (bp)   \n14                           Transcript end (bp)   \n15                Transcription start site (TSS)   \n16    Transcript length (including UTRs and CDS)   \n17                Transcript support level (TSL)   \n18                      GENCODE basic annotation   \n19                             APPRIS annotation   \n20                       RefSeq match transcript   \n21                                     Gene name   \n22                           Source of gene name   \n23                               Transcript name   \n24                     Source of transcript name   \n25                              Transcript count   \n26                             Gene % GC content   \n27                                     Gene type   \n28                               Transcript type   \n29                                 Source (gene)   \n...                                          ...   \n3321                                               \n3322                                               \n3323                                               \n3324                                               \n3325                                               \n3326                                               \n3327                                               \n3328                                               \n3329                     CDS start (within cDNA)   \n3330                       CDS end (within cDNA)   \n3331                                5' UTR start   \n3332                                  5' UTR end   \n3333                                3' UTR start   \n3334                                  3' UTR end   \n3335                                               \n3336                                               \n3337                                               \n3338                                               \n3339                                               \n3340                                               \n3341                                 start phase   \n3342                                   end phase   \n3343                           cDNA coding start   \n3344                             cDNA coding end   \n3345                        Genomic coding start   \n3346                          Genomic coding end   \n3347                                               \n3348                                  CDS Length   \n3349                                   CDS start   \n3350                                     CDS end   \n\n                                            description                dataset  \n0                                 Stable ID of the Gene  hsapiens_gene_ensembl  \n1                      Versionned stable ID of the Gene  hsapiens_gene_ensembl  \n2                           Stable ID of the Transcript  hsapiens_gene_ensembl  \n3                Versionned stable ID of the Transcript  hsapiens_gene_ensembl  \n4                                                        hsapiens_gene_ensembl  \n5                                                        hsapiens_gene_ensembl  \n6                                                        hsapiens_gene_ensembl  \n7                                                        hsapiens_gene_ensembl  \n8                              Chromosome/scaffold name  hsapiens_gene_ensembl  \n9     Start Coordinate of the gene in chromosomal co...  hsapiens_gene_ensembl  \n10    End Coordinate of the gene in chromosomal coor...  hsapiens_gene_ensembl  \n11         Orientation of genes on chromosome/scaffold.  hsapiens_gene_ensembl  \n12       Band of chromosome/scaffold where genes occur.  hsapiens_gene_ensembl  \n13    Start Coordinate of the transcript in chromoso...  hsapiens_gene_ensembl  \n14    End Coordinate of the gene in chromosomal coor...  hsapiens_gene_ensembl  \n15                                                       hsapiens_gene_ensembl  \n16                                                       hsapiens_gene_ensembl  \n17                                                       hsapiens_gene_ensembl  \n18                                                       hsapiens_gene_ensembl  \n19                                                       hsapiens_gene_ensembl  \n20                                                       hsapiens_gene_ensembl  \n21                                                       hsapiens_gene_ensembl  \n22                                                       hsapiens_gene_ensembl  \n23                                                       hsapiens_gene_ensembl  \n24                                                       hsapiens_gene_ensembl  \n25                                                       hsapiens_gene_ensembl  \n26                           Percentage gene GC content  hsapiens_gene_ensembl  \n27                                                       hsapiens_gene_ensembl  \n28                                                       hsapiens_gene_ensembl  \n29                                                       hsapiens_gene_ensembl  \n...                                                 ...                    ...  \n3321                                                     hsapiens_gene_ensembl  \n3322                                                     hsapiens_gene_ensembl  \n3323                                                     hsapiens_gene_ensembl  \n3324                                                     hsapiens_gene_ensembl  \n3325                                                     hsapiens_gene_ensembl  \n3326                                                     hsapiens_gene_ensembl  \n3327                                                     hsapiens_gene_ensembl  \n3328                                                     hsapiens_gene_ensembl  \n3329                                                     hsapiens_gene_ensembl  \n3330                                                     hsapiens_gene_ensembl  \n3331                                                     hsapiens_gene_ensembl  \n3332                                                     hsapiens_gene_ensembl  \n3333                                                     hsapiens_gene_ensembl  \n3334                                                     hsapiens_gene_ensembl  \n3335                                                     hsapiens_gene_ensembl  \n3336                                                     hsapiens_gene_ensembl  \n3337                                                     hsapiens_gene_ensembl  \n3338                                                     hsapiens_gene_ensembl  \n3339                                                     hsapiens_gene_ensembl  \n3340                                                     hsapiens_gene_ensembl  \n3341                                                     hsapiens_gene_ensembl  \n3342                                                     hsapiens_gene_ensembl  \n3343                                                     hsapiens_gene_ensembl  \n3344                                                     hsapiens_gene_ensembl  \n3345                                                     hsapiens_gene_ensembl  \n3346                                                     hsapiens_gene_ensembl  \n3347                                                     hsapiens_gene_ensembl  \n3348                                                     hsapiens_gene_ensembl  \n3349                                                     hsapiens_gene_ensembl  \n3350                                                     hsapiens_gene_ensembl  \n\n[3351 rows x 4 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>name</th>\n      <th>display_name</th>\n      <th>description</th>\n      <th>dataset</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ensembl_gene_id</td>\n      <td>Gene stable ID</td>\n      <td>Stable ID of the Gene</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>ensembl_gene_id_version</td>\n      <td>Gene stable ID version</td>\n      <td>Versionned stable ID of the Gene</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ensembl_transcript_id</td>\n      <td>Transcript stable ID</td>\n      <td>Stable ID of the Transcript</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ensembl_transcript_id_version</td>\n      <td>Transcript stable ID version</td>\n      <td>Versionned stable ID of the Transcript</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ensembl_peptide_id</td>\n      <td>Protein stable ID</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ensembl_peptide_id_version</td>\n      <td>Protein stable ID version</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>ensembl_exon_id</td>\n      <td>Exon stable ID</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>description</td>\n      <td>Gene description</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>chromosome_name</td>\n      <td>Chromosome/scaffold name</td>\n      <td>Chromosome/scaffold name</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>start_position</td>\n      <td>Gene start (bp)</td>\n      <td>Start Coordinate of the gene in chromosomal co...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>end_position</td>\n      <td>Gene end (bp)</td>\n      <td>End Coordinate of the gene in chromosomal coor...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>strand</td>\n      <td>Strand</td>\n      <td>Orientation of genes on chromosome/scaffold.</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>band</td>\n      <td>Karyotype band</td>\n      <td>Band of chromosome/scaffold where genes occur.</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>transcript_start</td>\n      <td>Transcript start (bp)</td>\n      <td>Start Coordinate of the transcript in chromoso...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>transcript_end</td>\n      <td>Transcript end (bp)</td>\n      <td>End Coordinate of the gene in chromosomal coor...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>transcription_start_site</td>\n      <td>Transcription start site (TSS)</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>transcript_length</td>\n      <td>Transcript length (including UTRs and CDS)</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>transcript_tsl</td>\n      <td>Transcript support level (TSL)</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>transcript_gencode_basic</td>\n      <td>GENCODE basic annotation</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>transcript_appris</td>\n      <td>APPRIS annotation</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>transcript_mane_select</td>\n      <td>RefSeq match transcript</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>external_gene_name</td>\n      <td>Gene name</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>external_gene_source</td>\n      <td>Source of gene name</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>external_transcript_name</td>\n      <td>Transcript name</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>external_transcript_source_name</td>\n      <td>Source of transcript name</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>transcript_count</td>\n      <td>Transcript count</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>26</th>\n      <td>percentage_gene_gc_content</td>\n      <td>Gene % GC content</td>\n      <td>Percentage gene GC content</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>27</th>\n      <td>gene_biotype</td>\n      <td>Gene type</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>28</th>\n      <td>transcript_biotype</td>\n      <td>Transcript type</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>29</th>\n      <td>source</td>\n      <td>Source (gene)</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>3321</th>\n      <td>sequence_transcript_biotype</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3322</th>\n      <td>sequence_transcript_chrom_strand</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3323</th>\n      <td>sequence_transcript_chrom_start</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3324</th>\n      <td>sequence_transcript_chrom_end</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3325</th>\n      <td>sequence_transcription_start_site</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3326</th>\n      <td>sequence_peptide_length</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3327</th>\n      <td>sequence_transcript_length</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3328</th>\n      <td>sequence_cdna_length</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3329</th>\n      <td>cdna_coding_start</td>\n      <td>CDS start (within cDNA)</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3330</th>\n      <td>cdna_coding_end</td>\n      <td>CDS end (within cDNA)</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3331</th>\n      <td>5_utr_start</td>\n      <td>5' UTR start</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3332</th>\n      <td>5_utr_end</td>\n      <td>5' UTR end</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3333</th>\n      <td>3_utr_start</td>\n      <td>3' UTR start</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3334</th>\n      <td>3_utr_end</td>\n      <td>3' UTR end</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3335</th>\n      <td>sequence_exon_stable_id</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3336</th>\n      <td>sequence_type</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3337</th>\n      <td>sequence_exon_chrom_start</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3338</th>\n      <td>sequence_exon_chrom_end</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3339</th>\n      <td>sequence_exon_chrom_strand</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3340</th>\n      <td>sequence_rank</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3341</th>\n      <td>sequence_phase</td>\n      <td>start phase</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3342</th>\n      <td>sequence_end_phase</td>\n      <td>end phase</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3343</th>\n      <td>sequence_cdna_coding_start</td>\n      <td>cDNA coding start</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3344</th>\n      <td>sequence_cdna_coding_end</td>\n      <td>cDNA coding end</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3345</th>\n      <td>sequence_genomic_coding_start</td>\n      <td>Genomic coding start</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3346</th>\n      <td>sequence_genomic_coding_end</td>\n      <td>Genomic coding end</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3347</th>\n      <td>sequence_constitutive</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3348</th>\n      <td>cds_length</td>\n      <td>CDS Length</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3349</th>\n      <td>cds_start</td>\n      <td>CDS start</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3350</th>\n      <td>cds_end</td>\n      <td>CDS end</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n  </tbody>\n</table>\n<p>3351 rows × 4 columns</p>\n</div>"
     },
     "metadata": {},
     "output_type": "execute_result",
     "execution_count": 3
    }
   ],
   "source": [
    "apybiomart.find_attributes()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### Filters  \n",
    "\n",
    "Filters, on the other hand, can be used to define a restriction on a query, such as retrieving only results from chromosome X. All available filters for a given dataset can be displayed using the `find_filters()` function, which by default will use the \"hsapiens_gene_ensembl\" dataset:  "
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [
    {
     "data": {
      "text/plain": "                                          name                    type  \\\n0                         link_so_mini_closure                    list   \n1                              link_go_closure                    text   \n2            link_ensembl_transcript_stable_id                    text   \n3                                      gene_id                    text   \n4                                transcript_id                    text   \n5                         link_ensembl_gene_id                    text   \n6                              chromosome_name                    text   \n7                                        start                    text   \n8                                          end                    text   \n9                                   band_start  drop_down_basic_filter   \n10                                    band_end  drop_down_basic_filter   \n11                                marker_start  drop_down_basic_filter   \n12                                  marker_end                           \n13                        hsapiens_encode.type                    text   \n14               hsapiens_encode.encode_region                           \n15                                      strand                    text   \n16                          chromosomal_region                    text   \n17                       id_list_xrefs_filters            boolean_list   \n18                 id_list_limit_xrefs_filters                 id_list   \n19                  id_list_microarray_filters            boolean_list   \n20            id_list_limit_microarray_filters                 id_list   \n21               transcript_count_greater_than                    text   \n22                  transcript_count_less_than                    text   \n23                                     biotype                    list   \n24                          transcript_biotype                    list   \n25                                      source                    list   \n26                           transcript_source                    list   \n27                              transcript_tsl                 boolean   \n28                    transcript_gencode_basic                 boolean   \n29                           transcript_appris                 boolean   \n30                                 mane_select                 boolean   \n31                       phenotype_description                    list   \n32                            phenotype_source                    list   \n33                                     go_clos                           \n34                                     go_name                           \n35                            go_evidence_code                    list   \n36                             homolog_filters            boolean_list   \n37  id_list_protein_domain_and_feature_filters            boolean_list   \n38        id_list_limit_protein_domain_filters                 id_list   \n39                  germ_line_variation_source                    list   \n40                    somatic_variation_source                    list   \n41                          with_validated_snp                 boolean   \n42                         so_consequence_name                    list   \n\n                                          description                dataset  \n0                                                      hsapiens_gene_ensembl  \n1                                                      hsapiens_gene_ensembl  \n2                                                      hsapiens_gene_ensembl  \n3                                                      hsapiens_gene_ensembl  \n4                                                      hsapiens_gene_ensembl  \n5   Filter to include genes with supplied list of ...  hsapiens_gene_ensembl  \n6                                                      hsapiens_gene_ensembl  \n7   Determine which base pair on the specified chr...  hsapiens_gene_ensembl  \n8   Determine which base pair on the specified chr...  hsapiens_gene_ensembl  \n9                                                      hsapiens_gene_ensembl  \n10                                                     hsapiens_gene_ensembl  \n11                                                     hsapiens_gene_ensembl  \n12                                                     hsapiens_gene_ensembl  \n13                                                     hsapiens_gene_ensembl  \n14                                                     hsapiens_gene_ensembl  \n15                                                     hsapiens_gene_ensembl  \n16  Limit to Genes within multiple comma separate ...  hsapiens_gene_ensembl  \n17                                                     hsapiens_gene_ensembl  \n18                                                     hsapiens_gene_ensembl  \n19                                                     hsapiens_gene_ensembl  \n20                                                     hsapiens_gene_ensembl  \n21  Limit to genes within given Transcript Count r...  hsapiens_gene_ensembl  \n22  Limit to genes within given Transcript Count r...  hsapiens_gene_ensembl  \n23                                                     hsapiens_gene_ensembl  \n24                                                     hsapiens_gene_ensembl  \n25                                                     hsapiens_gene_ensembl  \n26                                                     hsapiens_gene_ensembl  \n27                                                     hsapiens_gene_ensembl  \n28                                                     hsapiens_gene_ensembl  \n29                                                     hsapiens_gene_ensembl  \n30                                                     hsapiens_gene_ensembl  \n31                                                     hsapiens_gene_ensembl  \n32                                                     hsapiens_gene_ensembl  \n33                                                     hsapiens_gene_ensembl  \n34                                                     hsapiens_gene_ensembl  \n35                                                     hsapiens_gene_ensembl  \n36                                                     hsapiens_gene_ensembl  \n37                                                     hsapiens_gene_ensembl  \n38                                                     hsapiens_gene_ensembl  \n39                                                     hsapiens_gene_ensembl  \n40                                                     hsapiens_gene_ensembl  \n41  Entries associated with variants that have bee...  hsapiens_gene_ensembl  \n42                                                     hsapiens_gene_ensembl  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>name</th>\n      <th>type</th>\n      <th>description</th>\n      <th>dataset</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>link_so_mini_closure</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>link_go_closure</td>\n      <td>text</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>link_ensembl_transcript_stable_id</td>\n      <td>text</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>gene_id</td>\n      <td>text</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>transcript_id</td>\n      <td>text</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>link_ensembl_gene_id</td>\n      <td>text</td>\n      <td>Filter to include genes with supplied list of ...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>chromosome_name</td>\n      <td>text</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>start</td>\n      <td>text</td>\n      <td>Determine which base pair on the specified chr...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>end</td>\n      <td>text</td>\n      <td>Determine which base pair on the specified chr...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>band_start</td>\n      <td>drop_down_basic_filter</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>band_end</td>\n      <td>drop_down_basic_filter</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>marker_start</td>\n      <td>drop_down_basic_filter</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>marker_end</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>hsapiens_encode.type</td>\n      <td>text</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>hsapiens_encode.encode_region</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>strand</td>\n      <td>text</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>chromosomal_region</td>\n      <td>text</td>\n      <td>Limit to Genes within multiple comma separate ...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>id_list_xrefs_filters</td>\n      <td>boolean_list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>id_list_limit_xrefs_filters</td>\n      <td>id_list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>id_list_microarray_filters</td>\n      <td>boolean_list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>id_list_limit_microarray_filters</td>\n      <td>id_list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>transcript_count_greater_than</td>\n      <td>text</td>\n      <td>Limit to genes within given Transcript Count r...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>transcript_count_less_than</td>\n      <td>text</td>\n      <td>Limit to genes within given Transcript Count r...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>biotype</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>transcript_biotype</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>source</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>26</th>\n      <td>transcript_source</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>27</th>\n      <td>transcript_tsl</td>\n      <td>boolean</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>28</th>\n      <td>transcript_gencode_basic</td>\n      <td>boolean</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>29</th>\n      <td>transcript_appris</td>\n      <td>boolean</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>30</th>\n      <td>mane_select</td>\n      <td>boolean</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>31</th>\n      <td>phenotype_description</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>32</th>\n      <td>phenotype_source</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>33</th>\n      <td>go_clos</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>34</th>\n      <td>go_name</td>\n      <td></td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>35</th>\n      <td>go_evidence_code</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>36</th>\n      <td>homolog_filters</td>\n      <td>boolean_list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>37</th>\n      <td>id_list_protein_domain_and_feature_filters</td>\n      <td>boolean_list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>38</th>\n      <td>id_list_limit_protein_domain_filters</td>\n      <td>id_list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>39</th>\n      <td>germ_line_variation_source</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>40</th>\n      <td>somatic_variation_source</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>41</th>\n      <td>with_validated_snp</td>\n      <td>boolean</td>\n      <td>Entries associated with variants that have bee...</td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n    <tr>\n      <th>42</th>\n      <td>so_consequence_name</td>\n      <td>list</td>\n      <td></td>\n      <td>hsapiens_gene_ensembl</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "metadata": {},
     "output_type": "execute_result",
     "execution_count": 4
    }
   ],
   "source": [
    "apybiomart.find_filters()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  },
  "kernelspec": {
   "name": "python3",
   "language": "python",
   "display_name": "Python 3"
  },
  "stem_cell": {
   "cell_type": "raw",
   "source": "",
   "metadata": {
    "pycharm": {
     "metadata": false
    }
   }
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "source": [],
    "metadata": {
     "collapsed": false
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}